larpandoracontent/LArObjects/LArDiscreteProbabilityVector.cc

```diff
@@ -0,0 +1,237 @@
+
+        return m*x+c;
+    }
+
+    throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename TX, typename TY>
+DiscreteProbabilityVector::DiscreteProbabilityData DiscreteProbabilityVector::InitialiseDiscreteProbabilityData(
+    InputData<TX, TY> inputData) const
+{
+    if (2 > inputData.size())
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    std::sort(inputData.begin(), inputData.end(), DiscreteProbabilityVector::SortInputDataByX<TX,TY>);
+
+    if (inputData.back().first - m_xUpperBound > std::numeric_limits<float>::epsilon())
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    const float normalisation(this->CalculateNormalisation(inputData));
+    float accumulationDatum(0.f);
+
+    DiscreteProbabilityData data;
+    for (size_t iDatum = 0; iDatum < inputData.size()-1; ++iDatum)
+    {
+        const float x(static_cast<float>(inputData.at(iDatum).first));
+        const float deltaX(static_cast<float>(inputData.at(iDatum+1).first)-x);
+        const float densityDatum(static_cast<float>(inputData.at(iDatum).second) / normalisation);
+        accumulationDatum+=densityDatum*(m_useWidths ? deltaX : 1.f);
+        data.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(x, densityDatum, accumulationDatum, deltaX));
+    }
+    const float x(static_cast<float>(inputData.back().first));
+    const float deltaX(m_xUpperBound-x);
+    const float densityDatum(static_cast<float>(inputData.back().second) / normalisation);
+    accumulationDatum+=densityDatum*(m_useWidths ? deltaX : 1.f);
+    data.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(x, densityDatum, accumulationDatum, deltaX));
+
+    return data;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+DiscreteProbabilityVector::DiscreteProbabilityData DiscreteProbabilityVector::ResampleDiscreteProbabilityData(
+    const DiscreteProbabilityVector &discreteProbabilityVector, const ResamplingPoints &resamplingPoints) const
+{
+    if (2 > resamplingPoints.size())
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    DiscreteProbabilityData resampledProbabilityData;
+
+    float prevCumulativeData(0.f);
+    for (size_t iSample = 0; iSample < resamplingPoints.size()-1; ++iSample)
+    {
+        const float xResampled(resamplingPoints.at(iSample));
+        const float deltaX(resamplingPoints.at(iSample+1)-xResampled);
+        const float cumulativeDatumResampled(discreteProbabilityVector.EvaluateCumulativeProbability(xResampled));
+        const float densityDatumResampled((cumulativeDatumResampled-prevCumulativeData)/(m_useWidths ? deltaX : 1.f));
```

Comment:
Can deltaX be zero here?
Spaces around operators.
(And again, for both comments, below.)

Reply:
I got the spaces.

Actually it's possible for it to be 0 if the inputs are malformed.  The same thing could in principle be true when dividing by the normalisation in an above function.  I've added checks for those cases.